### ISSUE TITLE
[QUESTION] Conditional validation in Model

### ISSUE THREAD

**Issue by securityMB** [TypedProject/tsed#1284]
Hello!

Does Ts.ED provide a way to perform a conditional validation of a model based on another field in the model? Consider the following example

```typescript
export enum Buyer {
  Person = "PERSON",
  Company = "COMPANY"
}

export class Invoice {
  @Required()
  @Enum(Buyer)
  buyer: Buyer;

  @Optional()
  // I'd like this field to be required only if buyer === 'COMPANY'
  companyData?: CompanyData

  @Optional()
  // ... and this field should be required only if buyer === 'PERSON'
  // something like:
  // @RequiredIf(o => o.buyer === 'PERSON')
  personData?: PersonData
}
```

For the record, it doesn't really matter for me if this requirement is reflected in JSON Schema (both fields can be optional), I'm only caring about validation of request.

**Comment by Romakita** [comment]
Hello @securityMB 

There is no decorator  to do  that, but I found this on Ajv documentation: https://ajv.js.org/json-schema.html#compound-keywords

```
{
  type: "object",
  if: {properties: {foo: {minimum: 10}}},
  then: {required: ["bar"]},
  else: {required: ["baz"]}
}
```
So, by using the Schema decorator you can try the following example:
```typescript
export enum Buyer {
  Person = "PERSON",
  Company = "COMPANY"
}

@Schema({
   if: {
      properties: {
         buyer:  {
            const: "PERSON"
        }
      }
   },
   then: {
      required: "personData"
   }
})
export class Invoice {
  @Required()
  @Enum(Buyer)
  buyer: Buyer;

  @Optional()
  // I'd like this field to be required only if buyer === 'COMPANY'
  companyData?: CompanyData

  @Optional()
  // ... and this field should be required only if buyer === 'PERSON'
  // something like:
  // @RequiredIf(o => o.buyer === 'PERSON')
  personData?: PersonData
}
```
There is no solution right now to run custom validator based on Arrow function. Ajv doesn't provide this feature or you have to use custom Keyword: https://tsed.io/tutorials/ajv.html#user-defined-keywords
But this solution is a bit to complex just for your usecase.

Tell me if the solution works :)
See you
Romain

**Comment by securityMB** [comment]
@Romakita, thank you, it works like a charm!

The only minor problem is that `JsonSchema` doesn't have properties `if`, `then` and `else` so TypeScript is complaining about types.

**Comment by Romakita** [comment]
Yes you have to wrap object with any type to works :)

**Comment by securityMB** [comment]
I decided to just extend `JsonSchemaObject` like this:

```typescript
declare module "@tsed/schema" {
  interface JsonSchemaObject {
    if?: Partial<JsonSchemaObject>;
    then?: Partial<JsonSchemaObject>;
    else?: Partial<JsonSchemaObject>;
  }
}
```

No more complains :)

**Comment by securityMB** [comment]
Closing as the issue is solved.

**Comment by NachtRitter** [comment]
Hello @Romakita 
I had the similar case in my project.
What do you think about adding new validation decorator, smth like `@RequiredIf()` which will cover this case?

**Comment by Romakita** [comment]
@NachtRitter Because Ajv doesn't provide a way to have to have an arrow function to validate something. 

```typescript
 @RequiredIf(o => o.buyer === 'PERSON')
```

This annotation is not possible with Ajv. So If I implement this decorator, I'll also need to throw an exception that looks like the required one from.

If prefer to implement a decorator to perform custom validation/or static method instead:

```typescript
@OnValidate(data => {
    if (data.buyer === 'PERSON') {
       throw new ValidationError(....)
   }
})
class MyModel {

}
```
it's more flexible and isn't specific to AJV.

**Comment by xCryzed** [comment]
> @NachtRitter Because Ajv doesn't provide a way to have to have an arrow function to validate something.
> 
> ```ts
>  @RequiredIf(o => o.buyer === 'PERSON')
> ```
> 
> This annotation is not possible with Ajv. So If I implement this decorator, I'll also need to throw an exception that looks like the required one from.
> 
> If prefer to implement a decorator to perform custom validation/or static method instead:
> 
> ```ts
> @OnValidate(data => {
>     if (data.buyer === 'PERSON') {
>        throw new ValidationError(....)
>    }
> })
> class MyModel {
> 
> }
> ```
> 
> it's more flexible and isn't specific to AJV.

I have a similar use case and I think that would be a good solution.

**Comment by Romakita** [comment]
I close this issue. This feature will be implemented with #1336
